src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,887 @@
+            // SPI 0 is used on Raspberry with ChipSelectLine 1
+            var settings = new SpiConnectionSettings(busId, chipSelectLine);
+            // 500K is the SPI communication with GoPiGo
+            settings.ClockFrequency = 500000;
+            // see http://tightdev.net/SpiDev_Doc.pdf
+            settings.Mode = SpiMode.Mode0;
+            settings.DataBitLength = 8;
+            // as the SPI is a static, checking if it has already be initialised
+            if (_spiDevice == null)
+            {
+                _spiDevice = new UnixSpiDevice(settings);
+            }
+            GoPiGo3Info = new GoPiGoInfo();
+            if (autoDetect == true)
+                try
+                {
+                    GoPiGo3Info.Manufacturer = GetManufacturer();
+                    GoPiGo3Info.Board = GetBoard();
+                    GoPiGo3Info.SoftwareVersion = GetFirmwareVersion();
+                    GoPiGo3Info.HardwareVersion = GetHardwareVersion();
+                    GoPiGo3Info.Id = GetIdHex();
+                }
+                catch (IOException ex)
+                {
+                    throw new IOException($"No SPI response. GoPiGo3 with address {SpiAddress} not connected.", ex);
+                }
+            if ((GoPiGo3Info.Manufacturer != ManufacturerName) || (GoPiGo3Info.Board != BoardName))
+                throw new IOException($"GoPiGo3 with address {SpiAddress} not connected.");
+            MotorGearRatio = DefaultMotorGearRatio;
+            EncoderTicksPerRotation = DefaultEncoderTicksPerRotation;
+            // Initialise the 2 Groove sensors
+            GrooveSensor = new GrooveSensor[2]
+            {
+                new GrooveSensor(GroovePort.Groove1),
+                new GrooveSensor(GroovePort.Groove2)
+            };
+        }
+
+        /// <summary>
+        /// Reset everything
+        /// </summary>
+        public void Dispose()
+        {
+            // Reset groove sensors
+            SetGrooveType(GroovePort.Both, GrooveSensorType.Custom);
+            SetGrooveMode(GroovePort.Both, GrooveInputOutput.InputDigital);
+            // Turn off the motors
+            SetMotorPower(MotorPort.Both, (byte)MotorSpeed.Float);
+            // Reset the motor limits
+            SetMotorLimits(MotorPort.Both, 0, 0);
+            // Turn off the servos
+            SetServo(ServoPort.Both, 0);
+            // Turn off the LEDs
+            SetLed((byte)GoPiGo3Led.LedEyeLeft + (byte)GoPiGo3Led.LedEyeRight + (byte)GoPiGo3Led.LedBlinkerLeft + (byte)GoPiGo3Led.LedBlinkerRight, Color.Black);
+        }
+
+        #endregion
+
+        #region SPI transfer
+
+        /// <summary>
+        /// Conduct a SPI transaction
+        /// </summary>
+        /// <param name="dataOut">A byte array to send.The length of the array will determine how many bytes are transferred.</param>
+        /// <returns>Returns an array of the bytes read.</returns>
+        public byte[] SpiTransferArray(byte[] dataOut)
```

Comment:
>SpiTransferArray [](start = 22, length = 16)

Why is this public? This should be private instead and used by the public surface area instead as this is very low-level, you want users to have a higher level API where your device binding extracts all of the implementation logic out.

Reply:
All those Spi functions are public because if you want to build your own sensors, you need to have access to them. The build in sensors are limited to a subset of all the possible sensors. Some sensors are using different pins, can be combined, etc. And the only real way to access to them is to the low level transfer functions. They already abstract. I agree that this one may be private. But I'll keep the other ones public